api/git/service.go

```diff
@@ -0,0 +1,155 @@
+package git
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+)
+
+var (
+	ErrIncorrectRepositoryURL = errors.New("Git repository could not be found, please ensure that the URL is correct.")
+	ErrAuthenticationFailure  = errors.New("Authentication failed, please ensure that the git credentials are correct.")
+	ErrRefNotFound            = errors.New("The target ref is not found")
+
+	REPOSITORY_CACHE_SIZE = 4
+	REPOSITORY_CACHE_TTL  = 30 * time.Minute
+)
+
+type option struct {
```

Comment:
I don't think it's a good idea to use the same datatype for each method. I.e., download has nothing to do with `extensions` and having it there would be confusing for a future reader.
If you'd like to reuse commonly used fields across different structs, maybe embedding would be worth while:
```
type FetchOptions struct {
 repositoryUrl string
	username      string
	password      string
	referenceName string
}

type CloneOptions struct {
 FetchOptions
  depth int
 }
 ```